lib/src/matchers.rs

```diff
@@ -121,43 +122,36 @@ impl Matcher for FilesMatcher {
 }
 
 pub struct PrefixMatcher {
-    prefixes: BTreeSet<RepoPath>,
     dirs: Dirs,
 }
 
 impl PrefixMatcher {
     pub fn new(prefixes: &[RepoPath]) -> Self {
-        let prefixes = prefixes.iter().cloned().collect();
         let mut dirs = Dirs::new();
-        for prefix in &prefixes {
-            dirs.add_dir(prefix);
-            if !prefix.is_root() {
-                dirs.add_file(prefix);
-            }
+        for prefix in prefixes {
+            let sub = dirs.add(prefix);
+            sub.is_dir = true;
+            sub.is_file = true;
         }
-        PrefixMatcher { prefixes, dirs }
+        PrefixMatcher { dirs }
     }
 }
 
 impl Matcher for PrefixMatcher {
     fn matches(&self, file: &RepoPath) -> bool {
-        let components = file.components();
-        // TODO: Make Dirs a trie instead, so this can just walk that trie.
-        for i in 0..components.len() + 1 {
-            let prefix = RepoPath::from_components(components[0..i].to_vec());
-            if self.prefixes.contains(&prefix) {
-                return true;
-            }
-        }
-        false
+        self.dirs.walk_to(file).any(|(sub, _)| sub.is_file)
     }
 
     fn visit(&self, dir: &RepoPath) -> Visit {
-        if self.matches(dir) {
-            Visit::AllRecursively
-        } else {
-            self.dirs.get_visit_sets(dir)
+        for (sub, tail_components) in self.dirs.walk_to(dir) {
+            if sub.is_file {
+                return Visit::AllRecursively;
+            }
```

Comment:
This looks a little weird because visiting files recursively doesn't make much sense. Maybe add a comment saying that we abuse the `Dirs` type a bit by pretending that all prefixes we add are files (if I understood correctly that that's what we do).

Reply:
Yep, added inline comment.

Reply:
Actually, I wonder if we should rename `is_file` to something like `exact` since it seems to match that functionality from hg's matchers. Oh, maybe `explicit` and `implicit` would make sense for `is_file` and `is_dir`?

I also wonder if this refactoring will help us solve #505. Maybe that's even what inspired this PR?

Reply:
`explicit`/`implicit` might be better, but I haven't considered that seriously. `Visit` dirs/files are built from `RepoPathTree` entries, so is_dir/is_file naming makes sense there.

> I also wonder if this refactoring will help us solve #505

I have no idea, but a `RepoPathTree` could be composed from matchers to walk explicit paths.